pyqtorch/utils.py

```diff
@@ -75,9 +105,20 @@ def is_normalized(state: Tensor, atol: float = ATOL) -> bool:
 
 def is_diag(H: Tensor) -> bool:
     """
-    Returns True if Hamiltonian H is diagonal.
+    Returns True if tensor H is diagonal.
+
+    Reference: https://stackoverflow.com/questions/43884189/check-if-a-large-matrix-is-diagonal-matrix-in-python
+
+    Arguments:
+        H: Input tensor.
+
+    Returns:
+        True if diagonal, else False.
     """
-    return len(torch.abs(torch.triu(H, diagonal=1)).to_sparse().coalesce().values()) == 0
+    m = H.shape[0]
+    p, q = H.stride()
+    offdiag_view = torch.as_strided(H[:, 1:], (m - 1, m), (p + q, q))
+    return torch.count_nonzero(offdiag_view) == 0
```

Comment:
Cool :)

Is this much faster than the previous version?

One thing that will have to be "fixed" is to account for floating point arithmetic errors, since you may end up with very small non-zero off-diagonal elements. I believe this is what caused the previous version to fail, and I think it will also fail in this case. Anyway, this can easily be done by replacing the `torch.count_nonzero` by some efficient check of greater than $\epsilon$.

Reply:
Yes, much faster when I tried. I just added a tolerance argument for such cases.